### PR TITLE
(PXP-9319): Guppy to support cardinality queries

### PR DESCRIPTION
**Pull request by ocshawn**
Jira Ticket: [PXP-9319](https://ctds-planx.atlassian.net/browse/PXP-9319) [PPS-69](https://ctds-planx.atlassian.net/browse/PPS-69)

### New Features
- Guppy to support cardinality queries
- Lint guppy
- add countFields to connected filter that uses cardinality queries

[PXP-9319]: https://ctds-planx.atlassian.net/browse/PXP-9319?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[PPS-69]: https://ctds-planx.atlassian.net/browse/PPS-69?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: doc/queries.md [hunk 1]
```diff
@@ -765,6 +766,45 @@ Result:
 }
 ```
 
+<a name="aggs-cardinality"></a>
+
+### 6. Cardinality Count Aggregation
```
Comment (mfshao) [inline doc/queries.md:771]: I think we should make a note this doesn't work with nested fields yet?

File: src/components/Utils/queries.js [hunk 2]
```diff
@@ -23,31 +23,52 @@ const histogramQueryStrForEachField = (field) => {
...
+      }
+      `);
+  }
+  return (`
+  ${splittedField} {
+    ${cardinalityQueryStrForEachField(splittedFieldArray.join('.'))}
+  }`);
+};
+
+const queryGuppyForAggs = (path, type, fields, countFields, gqlFilter, acc) => {
```
Comment (mfshao) [inline src/components/Utils/queries.js:43]:
```suggestion
const queryGuppyForAggs = (path, type, fields, cardinalityFields, gqlFilter, acc) => {
```

I think this renaming can better reflect the purpose of this field

Also I think we should make it defaults to `{}` to make it optional for this function

**Review by mfshao — COMMENTED** [review]
overall I think it LGTM, you should also bump up the version number in package.json since you will need to make a new NPM release after merging this

**Review by mfshao — APPROVED** [review]
👍

File: src/server/resolvers.js [hunk 3]
```diff
@@ -123,6 +123,27 @@ const textHistogramResolver = async (parent, args, context) => {
...
+/**
+ * This resolver is for Cardinality.
+ * It inherits arguments from its parent,
+ * and uses "field" from parent and args "precision_threshold" to get the cardinality count
+ * @param {object} parent
+ * @param {object} args
+ */
+const cardinalityResolver = async (parent, args) => {
+  log.debug('[resolver.cardinalityResolver] args', args);
+  log.debug('[resolver.cardinalityResolver] parent', parent);
```
Comment (alilahrime) [inline src/server/resolvers.js:135]: dont think ya meant to leave these debug logs in

File: src/server/es/filter.js [hunk 4]
```diff
@@ -19,16 +19,16 @@ const fromPathToNode = (esInstance, esIndex, path) => {
 };
 
 const mergeRangeOperations = (a, b) => {
-  let merged = Object.assign({}, a, b);
+  const merged = { ...a, ...b };
 
-  Object.keys(merged).forEach(function(key) {
+  Object.keys(merged).forEach((key) => {
     if (typeof merged[key] === 'object' && merged[key] !== null) {
```
Comment (alilahrime) [inline src/server/es/filter.js:25]: `if (typeof merged[key] === 'object' && merged[key]) {`

File: src/components/Utils/queries.js [hunk 5]
```diff
@@ -23,31 +23,52 @@ const histogramQueryStrForEachField = (field) => {
...
+  }
+  return (`
+  ${splittedField} {
+    ${cardinalityQueryStrForEachField(splittedFieldArray.join('.'))}
+  }`);
+};
+
+const queryGuppyForAggs = (path, type, fields, cardinalityFields = [], gqlFilter, acc) => {
   let accessibility = acc;
   if (accessibility !== 'all' && accessibility !== 'accessible' && accessibility !== 'unaccessible') {
```
Comment (alilahrime) [inline src/components/Utils/queries.js:45]: `if (!['all', 'accessible', 'unaccessible'].includes(accessible)) {`

**Pull request merged**